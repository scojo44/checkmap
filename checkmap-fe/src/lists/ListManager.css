.ListManager ul {
  list-style-type: none;
}
.ListManager .list-row {
  display: flex;
  justify-content: space-between;
  margin-bottom: .5rem;
}
.ListManager .list-name {
  flex: 2;
}
.ListManager .list-button {
  flex: 1;
}
