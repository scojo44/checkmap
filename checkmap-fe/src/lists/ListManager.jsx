import React, {useContext, useState} from 'react'
import {useNavigate} from 'react-router-dom';
import CheckMapAPI from '../api';
import UserContext from '../UserContext'
import EditListForm from '../lists/EditListForm';
import './ListManager.css'

export default function ListManager(props) {
  const [editingListID, setEditingListID] = useState();
  const [lastDeletedListID, setLastDeletedListID] = useState();
  const navigate = useNavigate();
  const {user, currentList, setCurrentList, showAlert} = useContext(UserContext);

  return (
    <section className='ListManager'>
    <h2>Your Lists</h2>
    <ul className="user-lists">
      {user.lists.sort((a,b) => a.id > b.id).map(list => (
        <li key={list.id}>
          <div className="list-row">
            <span className="list-name" style={{fontWeight: editingListID === list.id? 'bold':''}}>{list.name}</span>
            <span className="list-button"><button onClick={() => showEditUI(list)}>Edit</button></span>
            <span className="list-button"><button onClick={() => switchList(list)} disabled={list === currentList}>Switch</button></span>
            <span className="list-button"><button onClick={() => deleteList(list)} disabled={list === currentList}>Delete</button></span>
          </div>
          {editingListID === list.id && <EditListForm list={list} update={updateList} cancel={hideEditUI}/>}
        </li>
      ))}
    </ul>
    <form>
      <button type="button" onClick={hideForm}>Close</button>
    </form>
    </section>
  );

  function switchList(list) { setCurrentList(list); }
  function showEditUI(list) { setEditingListID(list.id); }
  function hideEditUI() { setEditingListID(0); }
  function hideForm() { navigate('/map'); }

  async function updateList(updatedList) {
    try {
      const list = await CheckMapAPI.updateList(editingListID, updatedList);
      const userList = user.lists.find(l => l.id === list.id);
      userList.name = updatedList.name;
      userList.description = updatedList.description;
      userList.color = updatedList.color;
      if(list.id === currentList.id) setCurrentList({...userList});
      hideEditUI();
    }
    catch(e) {
      showAlert('error', 'Error creating list:' + e);
    }
  }

  async function deleteList(list) {
    try {
      const deletedListID = await CheckMapAPI.deleteList(list.id);
      const removeIndex = user.lists.findIndex(l => l.id === deletedListID);
      user.lists.splice(removeIndex, 1);
      setLastDeletedListID(deletedListID);
  }
    catch(e) {
      showAlert('error', 'Error creating list:' + e);
    }
  }
}
