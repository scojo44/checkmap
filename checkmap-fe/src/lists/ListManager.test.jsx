import {render} from '@testing-library/react'
import {MemoryRouter} from 'react-router-dom'
import UserContext from '../UserContext'
import ListManager from './ListManager'

const list = {
  id: 1,
  name: 'Test List 1a',
  description: 'This is a test list.',
  color: '#0000FF',
  regionType: 'State',
  ownerName: 'u1',
  states: [],
  counties: []
}
const user = {
  username: 'testuser',
  imageURL: 'u1.jpeg',
  role: 'User',
  lists: [list]
}
const context = {
  user,
  currentList: list,
  setCurrentList: x => x,
  showAlert: x => x
};

describe('ListManager Tests', () => {
  it('Renders without crashing', () => {
    render(
      <MemoryRouter>
        <UserContext.Provider value={context}>
          <ListManager />
        </UserContext.Provider>
      </MemoryRouter>
    );
  });

  it('Matches snapshot', () => {
    const {asFragment} = render(
      <MemoryRouter>
        <UserContext.Provider value={context}>
          <ListManager />
        </UserContext.Provider>
      </MemoryRouter>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
