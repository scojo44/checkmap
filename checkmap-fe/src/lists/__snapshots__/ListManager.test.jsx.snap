// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ListManager Tests > Matches snapshot 1`] = `
<DocumentFragment>
  <section
    class="ListManager"
  >
    <h2>
      Your Lists
    </h2>
    <ul
      class="user-lists"
    >
      <li>
        <div
          class="list-row"
        >
          <span
            class="list-name"
          >
            Test List 1a
          </span>
          <span
            class="list-button"
          >
            <button>
              Edit
            </button>
          </span>
          <span
            class="list-button"
          >
            <button
              disabled=""
            >
              Switch
            </button>
          </span>
          <span
            class="list-button"
          >
            <button
              disabled=""
            >
              Delete
            </button>
          </span>
        </div>
      </li>
    </ul>
    <form>
      <button
        type="button"
      >
        Close
      </button>
    </form>
  </section>
</DocumentFragment>
`;
